### ISSUE TITLE
portfinder@1.0.22 breaks ember serve

### ISSUE THREAD

**Issue by jelhan** [ember-cli/ember-cli#8794]
`portfinder` just released a `1.0.22` version. It breaks `ember serve` in two ways:

1. As soon as one upgrades to `portfinder@1.0.22` (or generate a new ember project / addon) `ember serve` always fails with

    > Port 4200 is already in use.
 
    This is also true for all other ports passed to `ember serve`, e.g. `ember serve --port 4300`. I'm not sure why this happens.

2. If port is > 40000 portfinder@1.0.22 throws:

    > Provided options.stopPort(40000is less than options.startPort (40001)

    This is caused by default port being reduced from `65535` to `40000`. This is reported upstream as https://github.com/http-party/node-portfinder/issues/85.

This will affect everybody who

- does not lock subdependencies
- upgrades ember-cli
- generates a new ember project (app or addon)

Ember CLI uses a version range that allows portfinder@1.0.22 for a few year. So nearly all active versions will be affected.

If using yarn, it's [resolutions feature](https://yarnpkg.com/lang/en/docs/selective-version-resolutions/) could be used as a quick work-a-round. To do so add a `resolutions` key to your `package.json`:

```
"resolutions": {
  "ember-cli/portfinder": "1.0.21"
}
```

Do not miss to run `yarn install` afterwards.

If you are not sure which version of `portfinder` is used in your project, run `yarn why portfinder`.

**Comment by kellyselden** [comment]
Thanks for digging into this. Hopefully they resolve it quickly.

**Comment by jelhan** [comment]
Another work-a-round was reported in a [question on StackOverflow](https://stackoverflow.com/questions/57537179/port-4200-or-any-other-port-i-try-is-already-in-use): `ember serve --port 0` In that case portfinder picks a random port, which seems to work fine.

**Comment by kellyselden** [comment]
Another workaround is to generate an older lockfile via `npm install --before 2019-08-16`.

**Comment by larryh** [comment]
Hi,

I ran into the same problem when upgrading from ember 3.11 to 3.12. You mentioned a fix for yarn - is there a similar fix for us npm users? 

I tried @kellyselden suggestion `npm install --before 2019-08-16` but that left the following entry in my package.lock.json file alone, so it still reads:

`
    "portfinder": {
      "version": "1.0.22",
      "resolved": "https://registry.npmjs.org/portfinder/-/portfinder-1.0.22.tgz",
      "integrity": "sha512-aZuwaz9ujJsyE8C5kurXAD8UmRxsJr+RtZWyQRvRk19Z2ri5uuHw5YS4tDBZrJlOS9Zw96uAbBuPb6W4wgvV5A==",
      "dev": true,
      "requires": {
        "async": "^1.5.2",
        "debug": "^2.2.0",
        "mkdirp": "0.5.x"
      },
`
I tried using `-- port 0` and it does generate and use a random port, but that means every time I fire up the server and get a new port number I have to change the bookmarks for "Ember Dev Server" and "Ember Tests" in my browser, in addition to CORS settings defined in my Rails back end. 

Which is obviously a pain in the butt.

**Comment by BarthesSimpson** [comment]
This fix is out for review: https://github.com/http-party/node-portfinder/pull/86

**Comment by jelhan** [comment]
> 1. As soon as one upgrades to `portfinder@1.0.22` (or generate a new ember project / addon) `ember serve` always fails with
>    > Port 4200 is already in use.
>    
>    
>    This is also true for all other ports passed to `ember serve`, e.g. `ember serve --port 4300`. I'm not sure why this happens.

This is caused by https://github.com/http-party/node-portfinder/issues/84. Ember CLI relies on sequential order of attempts. Especially if not called with `--port 0` it asserts that the default port `4200` or the port defined by `--port` is free. If not it throws the "Port ??? is already in use." error.

https://github.com/ember-cli/ember-cli/blob/a07650e8465fc8abe35a4a6f269b0f88d047ec58/lib/commands/serve.js#L114-L118

Actually the code here is less than optimal. We use `getPort` method to verify if a specific port is free but don't specify a `stopPort`. Therefore it searches for an open port between `commandOptions.port` and it's default highest port (`40000` for `1.0.22` and `65535` for older versions). But we are not interested in any other port than `commandOptions.port` (unless that one is `0`).

Should be refactored to something like

```js
let options = {
  host: commandOptions.host,
};

// if a specific port is given, we want to use that one or fail
if (commandOptions.port !== 0) {
  options.port: commandOptions.port,
  options.stopPort: commandOptions.port,
}

try {
  let foundPort = await getPort({ port: commandOptions.port, stopPort: commandOptions.port })
} catch (err) {
  let message = commandOptions.port !== 0 ?
    `Port ${commandOptions.port} is already in use.` : 
    `All ports above ${commandOptions.port} are already in use.`;
  return Promise.reject(new SilentError(message));
}
```

I'm not even sure if we need `getPort()` for the scenario `commandOptions.port !== 0` at all. It may be fine just to try to bind to the given port and fail with a meaningful error if it's already in use. That would also prevent us from race conditions. In current implementation a port might be free when calling `PortFinder.getPort()` but might not be anymore when actually trying to bin express server to it.

**Comment by imkathir** [comment]
I've encountered this as well.

**Comment by rwjblue** [comment]
This has been somewhat mitigated (@eriktrom published a 1.0.23 with the code from 1.0.21 while we work through the failures here).

**Comment by eriktrom** [comment]
> This has been somewhat mitigated (@eriktrom published a 1.0.23 with the code from 1.0.21 while we work through the failures here).

yes, more precisely, 1.0.23 **is** 1.0.21, exactly

(context: a good idea from a contributor had a couple issues, sorry all. Moving forward, we will be much more strict in releasing any changes)

**Comment by locks** [comment]
We are currently on 1.0.25, so I think it's safe to say the PR linked just before addressed this problem? Thanks for the discussion everyone!
